workflow/controller/controller.go

```diff
@@ -1033,6 +1033,10 @@ func (wfc *WorkflowController) newPodInformer(ctx context.Context) cache.SharedI
 					diff.LogChanges(oldPod, newPod)
 					return
 				}
+				// Skip initial deadlineExceed process, informer will receive another update after pod fully terminated
+				if newPod.Status.Reason == common.ErrDeadlineExceeded && oldPod.Status.Reason != common.ErrDeadlineExceeded {
```

Comment:
Added fix archiveLog annotation problem in deadlineExceed event for emissary
(because of network delays and queueing, the same problem may occur in pns/k8sapi as well, this fixes that)

![image](https://user-images.githubusercontent.com/20961507/142740706-841e6c61-94ee-4864-8157-301f709c417e.png)

Reply:
I think this is the wrong place for this. I can see new bugs appear in future. Whyt  happens if we pass the pod to be assessed, but then ignore it if deadline exceeded?

Reply:
Here are the approaches I have in mind:

1. use delayed pod clean up, 
after **main** container exit, pod is queue for delete straightaway
using delayed cleanup queue allows the informer to receive event from **wait** contianer exit
instead of triggering pod delete and informer never receive event from **wait** contianer exit

However, this introduces a new problem.
Even though the informer receive an event from **wait** container exit,
the workflow still won't get processed (won't update output log to status) due to condition `wf.Labels[common.LabelKeyCompleted] == "true"` (this is labeled from the **main** container exit)

Don't have a further idea on how to solve following this approach

2. skip process pod when main container exit due deadlineExceed
since there's always at least 2 pod update event to informer (1 from **main** container exit, 1 from **wait** container exit)
we skip the initial update (from reason Empty to reason deadlineExceed)
allow the wait container to finished and then have the controller annotate log, label complete

Reply:
> pod is queue for cleanup queue straightaway,

I thought we'd changed this to always delay clean-up by a short time (maybe 10s) by default. If that's not the case, then that solution is acceptable.

Reply:
> I thought we'd changed this to always delay clean-up by a short time (maybe 10s) by default. If that's not the case, then that solution is acceptable.

we have delay queue but don't have this in case of deadlineExceed

I tried added deadlineExceed in delayed queue 
but what would happen is the **main** container exit informer update labels the workflow as complete
and **wait** container exit informer update won't get processed because workflow is complete

At the end of the day, we still rely on a similar mechanism to understand when it is the first deadlineExceed and if it should get processed with the current level of information from pod.

In this case, is it good enough to approve and merge?

Reply:
Maybe there is another pod which also updates the workflow, or there is rsync (every 20m). This cannot prevent those edge cases, so we get runtime not-deterministic bugs. So we need something more.

Reply:
make sense, let me keep thinking

Reply:
A late hello in 2022.

The solution is now changed to use a delayed enqueue instead of ignoring the first deadline informer message.

The deadlineExceed from informer will add to queue after 1s and is not rate-limited.
I wonder if this is ok, or we should have another queue that delayed the enqueue to ratelimitedwfQueue (thinking about complexity here)

Reply:
No. Think about a workflow this 1000 pods. It'll thunder the controller. We can't have that. Instead, I think it would be fine for there to always be a delay between queuing a workflow, and reconcilliation. 

Is there a way we can update the rate limiter to always delay by 1s? That'd be fine.